cmd/thanos/compact.go

```diff
@@ -223,6 +235,60 @@ func runCompact(
 		}, func(error) {
 			cancel()
 		})
+
+		g.Add(func() error {
+			if retention.Seconds() == 0 {
+				level.Info(logger).Log("msg", "retention is disabled")
+
+				<-ctx.Done()
+			}
+
+			retentionFunc := func() error {
+				//Wait for compaction job to finish
+				mutex.Lock()
+				defer mutex.Unlock()
+
+				level.Info(logger).Log("msg", "start retention", "retention", retention)
+				err := bkt.Iter(ctx, "", func(name string) error {
+					id, ok := block.IsBlockDir(name)
+					if !ok {
+						return nil
+					}
+					m, err := block.DownloadMeta(ctx, logger, bkt, id)
+					if err != nil {
+						return errors.Wrap(err, "download metadata")
+					}
+
+					maxTime := time.Unix(m.MaxTime/1000, 0)
+					if time.Now().After(maxTime.Add(retention)) {
+						level.Info(logger).Log("msg", "deleting block", "id", id, "maxTime", maxTime.String())
+						if err := block.Delete(ctx, bkt, id); err != nil {
+							return errors.Wrap(err, "delete block")
+						}
+					}
+
+					return nil
+				})
+
+				if err != nil {
+					return errors.Wrap(err, "retention")
+				}
+
+				level.Info(logger).Log("msg", "retention done", "retention", retention)
+				return nil
+			}
+
+			if !wait {
+				return retentionFunc()
+			}
+
+			return runutil.Repeat(2*time.Hour, ctx.Done(), func() error {
+				return retentionFunc()
+			})
+		}, func(error) {
+			cancel()
```

Comment:
For future: we should never reuse context Timeouts for multiple run.Groups (: It should have it's own

Reply:
Why the case? As it is used only for closing?

Reply:
Yes. And it should not interact with other `run.Group`, as this function should close only the one it is tied to